pkg/cloud/rgraph/algo/parallel_queue.go

```diff
@@ -108,16 +108,18 @@ const (
 	stateDone
 )
 
-// Add an item to the queue. This method is threadsafe within op() and
-// can be called during Run(). It is NOT safe to call Add() from
-// a different, unassociated thread.
+// Add an item to the queue. This method is thread safe within op() and can be
+// called during Run(). It is NOT safe to call Add() from a different,
+// unassociated thread. This method returns an error when queue state is done.
 //
-// Calling Add(item) from the op() given to Run() guarantees that item
-// will be processed.
-func (q *ParallelQueue[T]) Add(item T) {
+// Calling Add(item) from the op() given to Run() guarantees that item will be
+// processed.
+func (q *ParallelQueue[T]) Add(item T) error {
 	q.lock.Lock()
 	defer q.lock.Unlock()
-
+	if q.state == stateDone {
+		return fmt.Errorf("Queue is done")
```

Comment:
If this is the only kind of error that is returned, I would rather this be a bool.

`return false`

Remember to change the docstring.

Reply:
I made it an error because then in place where it is used (like parallel-executor) we can print why the Add operation did not succeed. With bool we know if the operation succeeded or not but we need to know the implementation details to correlate it with the fact that the queue has finished and will not process more events. And in parallel executor we are aborting the whole operation not only this one action that did not succeed. So it is not obvious why we are doing that based on returned bool value. 

`Remember to change the docstring.`
I added the line in this method comments that it will return error when queue is done.

Reply:
But there is only one way it can currently fail -- your error message does not give more information than the bool.

Reply:
yes, right now it can fail only due to one reason but this is an implementation detail, right? 
There are 3 states here. 
1. queue is done - nothing will be process
2. queue is running - the action will be processed 100%
3. The queue is not started - the action will be process when it starts (eventually)

Then returning the bool is not clear what it means. Will the action be processed or not? In case of error we are signaling that sth is wrong with the queue and the client should not use it, and on the other hand it is ok to return nil for other cases because the queue is in a good state.
On the other hand, returning true when the queue is in NotStarted state is not that obvious for me. I'm not clear what it means and that's why I decided to go with the error. Because the message is more clear to me.  

Reply:
I don't quite follow the reasoning here:

You can state unambiguously that Add() returns False if the Queue is done.

1. It's not an error to Add() to a queue that hasn't started.
1. By using an `error` type, the user of this API needs to treat the return as potentially either a DONE situation OR maybe some other error. But in this case, Done is the only thing that can happen.
1. Your error is static (i.e. it's just one kind of error that is being returned). It does not convey more information.

Reply:
done